applications/asset_tracker/src/env_sensors/env_sensors.c

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <zephyr.h>
+#include <stdio.h>
+#include <string.h>
+#include <sensor.h>
+#include <spinlock.h>
+#include "env_sensors.h"
+
+struct env_sensor {
+	env_sensor_data_t sensor;
+	enum sensor_channel channel;
+	u8_t *dev_name;
+	struct device *dev;
+	struct k_spinlock lock;
+};
+
+static struct env_sensor temp_sensor = {
+	.sensor =  {
+		.type = ENV_SENSOR_TEMPERATURE,
+		.value = 0,
+	},
+	.channel = SENSOR_CHAN_AMBIENT_TEMP,
+	.dev_name = CONFIG_TEMP_DEV_NAME
+};
+
+static struct env_sensor humid_sensor = {
+	.sensor =  {
+		.type = ENV_SENSOR_HUMIDITY,
+		.value = 0,
+	},
+	.channel = SENSOR_CHAN_HUMIDITY,
+	.dev_name = CONFIG_TEMP_DEV_NAME
+};
+
+static struct env_sensor pressure_sensor = {
+	.sensor =  {
+		.type = ENV_SENSOR_AIR_PRESSURE,
+		.value = 0,
+	},
+	.channel = SENSOR_CHAN_PRESS,
+	.dev_name = CONFIG_TEMP_DEV_NAME
+};
+
+/* Array containg environment sensors available on the board. */
+static struct env_sensor *env_sensors[] = {
+	&temp_sensor,
+	&humid_sensor,
+	&pressure_sensor
+};
+
+static struct k_delayed_work env_sensors_poller;
+
+static void env_sensors_poll_fn(struct k_work *work)
+{
+	int num_sensors = ARRAY_SIZE(env_sensors);
+	struct sensor_value data[num_sensors];
+
+	int err;
+
+	if (IS_ENABLED(CONFIG_BME680)) {
+		err = sensor_sample_fetch_chan(env_sensors[0]->dev,
+					SENSOR_CHAN_ALL);
+		if (err) {
+			printk("Failed to fetch data from %s, error: %d\n",
+				env_sensors[0]->dev_name, err);
+		}
+	}
+	for (int i = 0; i < num_sensors; i++) {
+		if (!(IS_ENABLED(CONFIG_BME680))) {
+			err = sensor_sample_fetch_chan(env_sensors[i]->dev,
+				env_sensors[i]->channel);
+			if (err) {
+				printk("Failed to fetch data from %s, error: %d\n",
+					env_sensors[i]->dev_name, err);
+			}
+		}
+		err = sensor_channel_get(env_sensors[i]->dev,
+			env_sensors[i]->channel, &data[i]);
+		if (err) {
+			printk("Failed to fetch data from %s, error: %d\n",
+				env_sensors[i]->dev_name, err);
+		} else {
+			k_spinlock_key_t key = k_spin_lock(&(env_sensors[i]->lock));
```

Comment:
I am not sure it's a good idea to block in the system workqueue.

Reply:
The idea was to avoid reading of sensor data while writing the value. Spinlock is the fastest way to do this.

Reply:
Yes I get the idea, but that becomes necessary when we are reading these values periodically into static memory, from which we also read via API. Why aren't we simply reading what we need when the API is called, and let the caller provide the buffer? So we get rid of the blocking task in the system workqueue and all spinlocks.

Reply:
The bsec library implementation use the same interface and this library lives its own life with polling of the sensors at a regular interval. This kind of implementation also makes it safe to use this interface from multiple contexts. It started out with an ambition to be a more generic environment sensing module, but at the moment it is only a part of the asset tracker.

Reply:
This is not blocking the system workqueue though, it's just disabling interrupts for a very brief time?
Maybe @pizi-nordic can comment on this particular issue.

Reply:
On uniprocessor system the spinlock is equivalent of irq_lock(), so this is really the fastest method available. The only part we should take care of is the IRQ latency, but the protected operation is short.

However I am not sure what we are protecting here. The `data[i]` is local to the function, so there is no simultaneous access possible. The `sensor_value_to_double()` uses only argument to calculate the result and the argument is not going to change during the call, so it looks safe too. So the only `env_sensors[i]->sensor.value` seems to be protected.
So if this is the case, the code looks 100% OK.

Side comment:
The FPU on the Cortex-M4/M33 supports only single precision. Which means that the double is the same as float and has 32-bits. And this means that assignment of double/float is atomic, so no lock is needed here.
However the above statement is correct only if hardware FPU is used. The software emulation will use double precision math and the assignment will not be atomic any more.

Reply:
@pizi-nordic yes it was to protect env_sensors[i]->sensor.value from beeing accessed when it is written.